src/main/webapp/app/grading-system/grading-key-overview/grading-key-overview.component.ts

```diff
@@ -42,35 +42,53 @@ export class GradingKeyOverviewComponent implements OnInit {
-                    if (gradeSteps.maxPoints !== undefined) {
-                        if (!this.isExam) {
-                            // calculate course max points based on exercises
-                            const course = this.courseCalculationService.getCourse(this.courseId!);
-                            const maxPoints = this.courseCalculationService.calculateTotalScores(course!.exercises!, course!).get(ScoreType.REACHABLE_POINTS);
-                            this.gradingSystemService.setGradePoints(this.gradeSteps, maxPoints!);
-                        } else {
-                            // for exams the max points filed should equal the total max points (otherwise exams can't be started)
-                            this.gradingSystemService.setGradePoints(this.gradeSteps, gradeSteps.maxPoints!);
-                        }
+        // Note: This component is used in multiple routes, so it can be lazy loaded. Also, courseId and examId can be
+        // found on different levels of hierarchy tree (on the same level or a parent or a grandparent, etc.).
+        this.courseId = Number(this.findParamInRouteHierarchy('courseId'));
+        const examIdParam = this.findParamInRouteHierarchy('examId');
+        if (examIdParam) {
+            this.examId = Number(examIdParam);
+            this.isExam = true;
+        }
+        this.gradingSystemService.findGradeSteps(this.courseId, this.examId).subscribe((gradeSteps) => {
+            if (gradeSteps) {
+                this.title = gradeSteps.title;
+                this.isBonus = gradeSteps.gradeType === GradeType.BONUS;
+                this.gradeSteps = this.gradingSystemService.sortGradeSteps(gradeSteps.gradeSteps);
+                if (gradeSteps.maxPoints !== undefined) {
+                    if (!this.isExam) {
+                        // calculate course max points based on exercises
+                        const course = this.courseCalculationService.getCourse(this.courseId!);
+                        const maxPoints = this.courseCalculationService.calculateTotalScores(course!.exercises!, course!).get(ScoreType.REACHABLE_POINTS);
+                        this.gradingSystemService.setGradePoints(this.gradeSteps, maxPoints!);
+                    } else {
+                        // for exams the max points filed should equal the total max points (otherwise exams can't be started)
+                        this.gradingSystemService.setGradePoints(this.gradeSteps, gradeSteps.maxPoints!);
                     }
                 }
-            });
-        });
-        this.route.parent?.parent?.queryParams.subscribe((queryParams) => {
-            this.studentGrade = queryParams['grade'];
+            }
         });
+
+        // Needed queryParam is available on this component so no need to traverse the hierarchy like params above.
+        this.studentGrade = this.route.snapshot.queryParams['grade'];
```

Comment:
The old code called subscribe() here and above, now you directly access the value. Do both ways work?

Reply:
Yes, since the component does not update params or queryParams dynamically after page load, it still works. Actually the important change was removing the `.parent?.parent?` part but I also used `snapshot` here to make it consistent with the `findParamInRouteHierarchy` usages above.